High Latency Protocol

hamishwillee opened the pull request:
@DonLakeFlyer This is draft docs for the High latency API following our discussion [here](https://github.com/mavlink/qgroundcontrol/pull/9101#issuecomment-704524299). 

Can you please let me know if these assumptions are correct (yes, no, maybe, don't know)
1. Vehicle pretty much _only_ emits HIGH_LATENCY2 and it acts as both heartbeat and as essential telemetry. 
2. Vehicle doesn't emit other messages.
3. Vehicle only emits HIGH_LATENCY2  message when commanded by MAV_CMD_CONTROL_HIGH_LATENCY 
4. QGC sends MAV_CMD_CONTROL_HIGH_LATENCY when an HL link becomes the primary. 
5. The implication of 4 is that the connection is unlike a normal LL connection where you start communicating on the same link by sending heartbeats and you know that the connection has fallen out when the heartbeat messages stop. In this case the link is only "theoretical" - QGC knows about it and can connect to it, but it isn't actually starting to monitor for HL2 messages until after sending that command. Presumably it still uses a timeout for when HL message not received for ages.

Is that all about right/can you confirm? Do you know what rate QGC expects for a HIGH_LATENCY2 timeout?

Note, I have marked `HIGH_LATENCY` as deprecated in favour of `HIGH_LATENCY2`: https://github.com/mavlink/mavlink/pull/1505

DonLakeFlyer commented:
"The GCS sends this to the vehicle to enable/disable sending information over the high latency link. This might enabled if the user explicitly sets the high latency link as primary, or if all low latency links are lost."

This needs some more explanation. The scenario is if there are multiple links to the vehicle and the primary/secondary link assignment changes. 

Examples:
* HL link is currently primary
* A new non HL link shows up
* This will cause that new link to become primary
* The HL link now becomes secondary and the telemetry on it will be disabled using `CONTROL_HIGH_LATENCY`

* There are two links
* Primary is non HL link
* Secondary is HL
* Comm lost on primary
* HL will move to primary and `CONTROL_HIGH_LATENCY` is sent to start telemetry

Not sure how you explain all of that in a reasonable way.

hamishwillee commented:
Thanks @DonLakeFlyer . Good points. I have updated explanation and added diagram to capture that - better? 

https://github.com/mavlink/mavlink-devguide/blob/b569f4fca9d35f4a40ef4bfe59a3f2a0ad1d8af2/en/services/high_latency.md

DonLakeFlyer commented:
Look fine
